blockchain/store_test.go

```diff
@@ -0,0 +1,269 @@
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:      validPartSet,
+			seenCommit: seenCommit1,
+		},
+
+		{
+			block:     nil,
+			wantPanic: true, // want a panic on saving nil block
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  4,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:     uncontiguousPartSet,
+			wantPanic: true, // and incomplete and uncontiguous parts
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:     incompletePartSet,
+			wantPanic: true, // incomplete parts
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:            validPartSet,
+			seenCommit:       seenCommit1,
+			eraseDBForCommit: true,
+			wantPanic:        true, // erase all content so that the commit will not be read
+		},
+	}
+
+	type quad struct {
+		block  *types.Block
+		commit *types.Commit
+		meta   *types.BlockMeta
+
+		seenCommit *types.Commit
+	}
+
+	for i, tuple := range tuples {
+		bs, db := freshBlockStore()
+		// SaveBlock
+		res, err, panicErr := doFn(func() (interface{}, error) {
+			bs.SaveBlock(tuple.block, tuple.parts, tuple.seenCommit)
+			if tuple.block == nil {
+				return nil, nil
+			}
+			bBlock := bs.LoadBlock(tuple.block.Height)
+			bBlockMeta := bs.LoadBlockMeta(tuple.block.Height)
+			bSeenCommit := bs.LoadSeenCommit(tuple.block.Height)
+			if tuple.eraseDBForCommit {
+				eraseDB(db)
+			}
+			bCommit := bs.LoadBlockCommit(tuple.block.Height - 1)
+			return &quad{block: bBlock, seenCommit: bSeenCommit, commit: bCommit, meta: bBlockMeta}, nil
+		})
+
+		if tuple.wantPanic {
+			if panicErr == nil {
+				t.Errorf("#%d: want a non-nil panic", i)
+			}
+			continue
+		}
+
+		if tuple.wantErr {
+			if err == nil {
+				t.Errorf("#%d: got nil error", i)
+			}
+			continue
+		}
+
+		assert.Nil(t, panicErr, "#%d: unexpected panic", i)
+		assert.Nil(t, err, "#%d: expecting a non-nil error", i)
+		qua, ok := res.(*quad)
+		if !ok || qua == nil {
+			t.Errorf("#%d: got nil quad back; gotType=%T", i, res)
+			continue
+		}
+	}
+}
+
+func doFn(fn func() (interface{}, error)) (res interface{}, err error, panicErr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			stack := make([]byte, 1024)
+			_ = runtime.Stack(stack, true)
+			panicErr = fmt.Errorf("%s", stack)
+		}
+	}()
+
+	res, err = fn()
+	return res, err, panicErr
+}
+
+func eraseDB(db db.DB) {
+	it := db.Iterator()
+	for {
```

Comment:
can't we just do:

```
for it.Next(){
 db.Delete(it.Key())
}
```

Reply:
For some reason that was initially buggy on my side and not erasing the last element sometime back, but that now looks functional, thank you for the update.